From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Olli Kokko <oko@iki.fi>
Date: Thu, 24 Sep 2026 17:16:32 +0300
Subject: [PATCH] Fix #1622: Double free on publish with QoS > 0 on
 SOCKET_ERROR

When QoS > 0, no copy is made when publish is started, and
storePublication takes ownership and nulls the payload of the Publish.
If publish fails with SOCKET_ERROR, the command is freed without
remembering to null the command's payload pointer, and thus the same
thing is freed again when the publications are removed.

Signed-off-by: Olli Kokko <oko@iki.fi>
---
 src/MQTTAsyncUtils.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/src/MQTTAsyncUtils.c b/src/MQTTAsyncUtils.c
index c084a11..de1ba90 100644
--- a/src/MQTTAsyncUtils.c
+++ b/src/MQTTAsyncUtils.c
@@ -1625,6 +1625,15 @@ static int MQTTAsync_processCommand(void)
 				command->client->connect = command->command;
 				MQTTAsync_startConnectRetry(command->client);
 			}
+			if (command->command.type == PUBLISH && command->command.details.pub.qos > 0)
+			{
+				/* For QoS 1 and 2, MQTTProtocol_startPublish has already stored the publication,
+				 * which took ownership of the topic and payload. They will be freed when the
+				 * publication is removed, so don't free them here as well.
+				 */
+				command->command.details.pub.destinationName = NULL;
+				command->command.details.pub.payload = NULL;
+			}
 			MQTTAsync_freeCommand(command);  /* free up the command if necessary */
 		}
 	}
